File: neetcode150/arrays_and_hashing/leetcode242.py
from collections import Counter


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        s_counter = Counter(s)

        for ch in t:
            if ch not in s_counter:
                return False

            s_counter[ch] -= 1

        return all([value == 0 for value in s_counter.values()])

File: js_algorithms/neetcode150/array_and_hashing/leetcode242.js
const generateCounter = require("../../utils/generateCounter");
const sum = require("../../utils/sum");

/**
 * 242. Valid Anagram
 *
 * Given two strings s and t, return true if t is an anagram of s, and false otherwise.
 *
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
 * typically using all the original letters exactly once.
 *
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
const isAnagram = function (s, t) {
  const sCounter = generateCounter(s.split(""));

  for (const ch of t) {
    const currentCount = sCounter.get(ch) ?? 0;
    if (currentCount === 0) {
      return false;
    }

    sCounter.set(ch, currentCount - 1);
  }

  return sum([...sCounter.values()]) === 0;
};

module.exports = { isAnagram };
